### PR TITLE
Add support for node labelled graphs

### PR DESCRIPTION
**Pull request by s1ck**

### CODE REVIEW

File: src/input/dotgraph.rs [hunk 1]
```diff
@@ -33,14 +33,70 @@ impl<Node: Idx, Label: Idx> InputCapabilities<Node> for DotGraphInput<Node, Labe
...
+        for label in Label::zero()..Label::new(label_count) {
+            offsets.push(total);
+            let freq = self.label_frequency.get(&label).unwrap_or(&0);
+            total += Label::new(*freq);
+        }
+
+        for (node, &label) in self.labels.iter().enumerate().take(node_count.index()) {
+            let offset = offsets[(label + Label::new(1)).index()];
+            nodes[offset.index()] = Node::new(node);
+            offsets[(label + Label::new(1)).index()] += Label::new(1);
```
Comment (adamnsch) [inline src/input/dotgraph.rs]: maybe just make var `offset` mut and add to that here instead?

File: src/input/dotgraph.rs [hunk 2]
```diff
@@ -33,14 +33,70 @@ impl<Node: Idx, Label: Idx> InputCapabilities<Node> for DotGraphInput<Node, Labe
...
+        let mut total = Label::zero();
+
+        for label in Label::zero()..Label::new(label_count) {
+            offsets.push(total);
+            let freq = self.label_frequency.get(&label).unwrap_or(&0);
+            total += Label::new(*freq);
+        }
+
+        for (node, &label) in self.labels.iter().enumerate().take(node_count.index()) {
+            let offset = offsets[(label + Label::new(1)).index()];
```
Comment (adamnsch) [inline src/input/dotgraph.rs]: Don't understand why we `+ Label::new(1)` here, guess I'm missing something

File: src/input/dotgraph.rs [hunk 3]
```diff
@@ -33,14 +33,70 @@ impl<Node: Idx, Label: Idx> InputCapabilities<Node> for DotGraphInput<Node, Labe
...
+impl<Node, Label> DotGraph<Node, Label>
+where
+    Node: Idx,
+    Label: Idx + Hash,
+{
+    fn node_count(&self) -> Node {
+        Node::new(self.labels.len())
+    }
+
+    pub(crate) fn label_count(&self) -> usize {
```
Comment (adamnsch) [inline src/input/dotgraph.rs]: Not sure I understand the logic of this one. Maybe a comment could be nice? :relaxed:

**Review by adamnsch — APPROVED** [review]
Impressive work :) Learned a lot about the lib reading this. Still some minor things I don't completely understand haha, but the code looks really good :+1:

File: src/input/dotgraph.rs [hunk 4]
```diff
@@ -1,41 +1,177 @@
...
+        let (node_count, used) = Node::parse(header);
+        header = &header[used + 1..];
+        let (edge_count, _) = Node::parse(header);
+
+        let mut labels = Vec::<Label>::with_capacity(node_count.index());
+        let mut edges = Vec::with_capacity(edge_count.index());
+
+        let mut max_degree = Node::zero();
+        let mut max_label = Label::zero();
+        let mut label_frequency = HashMap::<Label, usize>::new();
```
[the remark this answers is not in the document]
Comment (s1ck) [reply src/input/dotgraph.rs:206]: Good point. We can pick a different hasher already as we don't need the security.

File: src/input/dotgraph.rs [hunk 5]
```diff
@@ -1,41 +1,177 @@
...
+            batch = &batch[used + 1..];
+            let (label, used) = Label::parse(batch);
+            batch = &batch[used + 1..];
+            let (degree, used) = Node::parse(batch);
+            batch = &batch[used + 1..];
+
+            labels.push(label);
+
+            if degree > max_degree {
+                max_degree = degree;
```
[the remark this answers is not in the document]
Comment (s1ck) [reply src/input/dotgraph.rs:228]: ok ^^

File: src/graph/csr.rs [hunk 6]
```diff
@@ -277,15 +277,15 @@ impl<Node: Idx> DirectedGraph<Node> for DirectedCsrGraph<Node> {
...
+        self.out_edges.targets(node)
     }
 
     fn in_degree(&self, node: Node) -> Node {
         self.in_edges.degree(node)
     }
 
     fn in_neighbors(&self, node: Node) -> &[Node] {
-        self.in_edges.neighbors(node)
+        self.in_edges.targets(node)
```
[the remark this answers is not in the document]
Comment (s1ck) [reply src/graph/csr.rs:288]: maybe we should the fields `csr`, `csr_in` and `csr_out` instead?

[the remark this answers is not in the document]
Comment (s1ck) [reply src/graph/csr.rs:288]: or `csr_out` and `csr_inc` because they have the same length :)

File: src/input/dotgraph.rs (hunk 1 above)
Comment (s1ck) [reply src/input/dotgraph.rs]: I think this is outdated. I replaced this with a multi-threaded impl.

File: src/input/dotgraph.rs (hunk 2 above)
Comment (s1ck) [reply src/input/dotgraph.rs]: Filling the `targets` array works based off the `offsets`. To know where we need to insert a target id, we get and increment the offset for that particular label and insert at the resulting index. For example, if our offsets are:

```
0 3 4 7
```

and we retrieve the index to insert into the targets array from those values by we'd end up with

```
3 4 7 7
```

In another part of the code, we correct this by using `rotate_right(1)` and `push(0)`.

In this impl I tried something else, we place the offsets one index to the right:

```
0 0 3 4 7
```

and when we retrieve the offset for e.g. index 0, we actually update the offset value for index 1, resulting in a correct offsets array after inserting into the targets array:

```
0 3 4 7 7
```

Basically saves us the `rotate_right`.

File: src/input/dotgraph.rs (hunk 3 above)
Comment (s1ck) [reply src/input/dotgraph.rs]: Mh... not sure what I thought while writing this. It should never be the case that the length of the map is larger than the maximum label. The other way around is true, e.g. you can have labels, `42` and `1337` and the `frequencies.len()` is `2`. We however need `1338` to correctly initialize the offsets for the label index csr. I simplified it.

File: src/graph/csr.rs [hunk 7]
```diff
@@ -50,46 +50,46 @@ impl Default for CsrLayout {
...
-    offsets: Box<[Node]>,
-    targets: Box<[Node]>,
+pub struct Csr<Index: Idx, T> {
+    offsets: Box<[Index]>,
+    targets: Box<[T]>,
 }
 
-impl<Node: Idx> Csr<Node> {
-    pub(crate) fn new(offsets: Box<[Node]>, targets: Box<[Node]>) -> Self {
+impl<Index: Idx, T> Csr<Index, T> {
```
Comment (knutwalker) [inline src/graph/csr.rs:58]: Could use a default type to signal the common case, e.g.

```suggestion
impl<Index: Idx, T> Csr<Index, T = Index> {
```

which would allow using `Csr<Node>` as an alias for `Csr<Node, Node>`.

File: src/input/dotgraph.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (s1ck) [reply src/input/dotgraph.rs]: I used https://github.com/cbreeden/fxhash instead

[the remark this answers is not in the document]
Comment (knutwalker) [reply src/input/dotgraph.rs]: [ahash](https://crates.io/crates/ahash) or [fxhash](https://crates.io/crates/fxhash) seem frequent candidates. [nohash](https://crates.io/crates/nohash-hasher) could also work, though the non-distributed nature of identifiers might lead to a lot of collisions if the bucket selections depends on diversity in the higher bits.
Using one of these looks like a candidate for a feature, though depending on the hasher as a new generic type (like HashMap already does) is probably more flexible.

**Comment by s1ck** [comment]
@adamnsch Thanks for the review. I'm not entirely sure how much of the extra info like label frequency, max label and the reverse index will eventually stay in the lib. I mainly use this in my subgraph project, where I also could create a graph with such extra infos. Here, a labelled graph would maybe just be defined by the extra `fn label(node: Node) -> Label` :thinking:

File: src/graph/csr.rs (hunk 7 above)
Comment (s1ck) [reply src/graph/csr.rs:58]: So nice! :heart: :crab:
